### DISCUSSION TITLE
search.refresh does not clear cache

### DISCUSSION THREAD

**Discussion by czmarcos** [algolia/instantsearch#5900] [Q&A] [answered]
Hi guys,

I'm using instantsearch js version @4.57.0 and I have an instant search application which is integrated with some features that alter the hits of the table (through backend/serverside). As I update an object, I tried to apply "search.refresh()" in order to refresh the hits, however it does not return the updated data, while looking at the logs applying a refresh with a big delay it always brings the old data as well.

I also tried to apply "search.clearCache()" and "search.client.clearCache()" as indicated on [this issue](https://github.com/algolia/instantsearch/issues/1050), however it is returning "search.clearCache is not a function".

It looks like it is a cache problem, as while I look at Algolia's dashboard the data is updated within few milliseconds. Does anyone has any possible guess how can I bring updated hits?

```
var search = instantsearch({
        searchClient: algoliasearch(
          'ABC',
          'abc'
        ),
        indexName: 'my-index',
        routing: true,
      });
```

```
const updateData = ()  => {
    // this function is just an example that would update some data through backend and then I want to refresh the hits
    search.clearCache();
    search.refresh();
}
```

thanks!

**Comment by Haroenv** [comment]
all refresh does is clear the cache and search again: https://github.com/algolia/instantsearch/blob/6abc8f7a5d7a84ccd2188300d65ee3aac321c82d/packages/instantsearch.js/src/lib/InstantSearch.ts#L834-L843

What I'm guessing is while the data updates quite fast in Algolia, it doesn't happen _instantly_. That means you probably do the table update, and refresh right away, before it's available in the search results.

For solving that, you can add `.wait()` to the operation you did changing the index before `search.refresh()`


Oh and `search.clearCache` doesn't exist, only `helper.clearCache` and `searchClient.clearCache`

**Comment by czmarcos** [comment, marked as answer]
@Haroenv thanks for your tips!

Got it, I tried a timeout of more than 60s and the refresh did not bring updated hits, but if I do a simple browser refresh whit almost none delay the new search data is updated fine.

Regarding using .wait() I guess that couldn't be used on my case, because the data is updated on backend (server side). Am I right?

Thanks!

**Reply by Haroenv** [reply]
I have an application that relies on refresh that I'm just now working on, and I definitely see refresh working. If the actual edit happens backend you can't use wait indeed, maybe you can use a socket or server sent event to let the client know to refresh, but that may be overkill. 

Do you see a request happen when you call `.refresh` or not? if not a reproduction would be interesting to see

**Reply by czmarcos** [reply]
got it!

yes, I use a timeout to delay 10s and then call refresh, I can confirm the function is called and return the search with full hit results (don't know why, but it returns twice). Tried to look if there is any object that indicates if the results is from cache or not or any other useful info, but did not find.

**Reply by Haroenv** [reply]
if a network request happened, the results are not from cache :)

**Reply by czmarcos** [reply]
hmm nice hint! :)

so search.refresh results were always from cache, as no network request happened and results returned.

thus search.refresh was not working as expected per documentation, so I tried to update every algolia's search/instantsearch versions and it worked fine now!

thanks a lot!
